## SNP检测和注释

单核苷酸多态性（Single Nucleotide Polymorphism，SNP），是指基因组中由单个核苷酸的变异所引起的DNA序列多态性，是基因组上多态性最高的遗传变异之一。SNP变异类型分为转换和颠换两种，同种类型碱基（嘌呤与嘌呤、嘧啶与嘧啶）之间的突变称为转换（Transition）；不同类型碱基（嘌呤与嘧啶）之间的突变称为颠换（Transversion）。一般转换比颠换更容易发生，所以转换/颠换（Ts/Tv）的比例一般大于1，具体比值和所测物种有关。


### SNP检测

利用GATK的Best Practices流程处理比对结果（BAM文件），利用GATK的Haplotyper方法进行SNP检测，过滤条件按照GATK推荐的参数进行，具体可见[https://software.broadinstitute.org/gatk/documentation/article.php?id=3225](https://software.broadinstitute.org/gatk/documentation/article.php?id=3225)。样品SNP统计结果见表\@ref(tab:snpstat-table)：


```{r get-snpstat-table}
snp_stat_file <- file.path(wgs_result_dir,"output","published","data", "04.snpIndel", "snp", "snp.stat.xls")
snp_stat <- read.delim(snp_stat_file,sep='\t',header=TRUE,check.names = FALSE)
snp_stat$num_homo <- snp_stat$num_homo_ref + snp_stat$num_homo_alt
tempdf <- snp_stat[,c("sample","num_snv","num_ts","num_tv","ts/tv","num_het","num_homo")]
colnames(tempdf) <- c("Sample ID","SNP Number","Transition","Transversion","Ts/Tv","Heterozygosity Number","Homozygosity Number")
tempdf <- tempdf %>% tbl_prettier(c(NA,0,0,0,2,0,0))
tempdf <- tempdf[str_order(tempdf$`Sample ID`, numeric = T),]
tempfn <- c(
"Sample ID：样品编号；",
"SNP Number：检测到的单核苷酸多态性位点的数量，表示材料与参考基因组之间的核苷酸变异；",
"Transition：转换的SNP数量；",
"Transversion：颠换的SNP数量；",
"Ts/Tv：转换型SNP（Transition）和颠换型SNP（Transversion）的比值；",
"Heterozygosity Number：杂合分型的SNP位点总数；",
"Homozygosity Number：纯合分型的SNP位点总数。"
)
```

```{r snpstat-table-html,eval=htmlfmt,results="asis"}
dt_caption("SNP数据统计表","snpstat-table")
datatable(tempdf, rownames=F,options = list(scrollX=T,columnDefs = list(list(className = 'dt-center',targets='_all'))))
dt_foot(tempfn)
```

```{r snpstat-table,eval=pdffmt}
tempdf[,1]=gsub("_","\\\\_",tempdf[,1])
kbl(row.names=FALSE, tempdf, "latex",caption="SNP数据统计表", booktabs = TRUE, longtable = TRUE, linesep=' ',col.names=linebreak(c("Sample ID","SNP\nNumber","Transition","Transversion","Ts/Tv","Heterozygosity\nNumber","Homozygosity\nNumber"),align="c"),escape=FALSE,align="c") %>%
kable_styling(repeat_header_text="(续)", latex_options=c("striped","scale_down","repeat_header"))%>%
kbl_foot(tempfn) %>% table_bline()
```

### SNP功能注释

采用SnpEff（Cingolani *et al.* 2012）程序结合本项目`r paste0(project_info[4],"（",project_info[5],"）")`基因组注释信息，对检测到的SNP进行功能注释，SnpEff会根据基因组的基因和功能区域的分布进行分析，对每个SNP所在的位置和功能进行统计，并对每个变异类型的功能进行统计。表\@ref(tab:snpannostat-table)为SNP位置分布信息统计表，表\@ref(tab:effectstat-table)为SNP功效信息统计表。

> *完整的SNP功能信息统计详见结题文件夹目录[data_release/01.vcf_filter](./data_release/01.vcf_filter)下的pop.snp_anno.xls文件*

```{r get-snpannostat-table}
snp_anno_stat_file <- file.path(wgs_result_dir,"output","published","data", "04.snpIndel", "snp","snp_anno.stat.xls")
snp_anno_stat <- read.delim(snp_anno_stat_file, sep='\t',header=T,check.names = F)
tempdf <- snp_anno_stat[,c("sampleID","start_lost","stop_lost","stop_gained","missense_variant","synonymous_variant","intergenic_region")]
colnames(tempdf) <- str_to_title(gsub("_"," ",colnames(tempdf)))
colnames(tempdf)[1] <- "Sample ID"
tempdf <- tempdf %>% tbl_prettier(c(NA,0,0,0,0,0,0))
tempdf <- tempdf[str_order(tempdf$`Sample ID`, numeric = T),]
tempfn <- c(
    "Sample ID：样品编号；",
    "Start Lost：由于SNP的突变导致启动子缺失的SNP位点个数；",
    "Stop Lost：由于SNP的突变导致终止子突变的SNP位点个数；",
    "Stop Gained：由于SNP的突变导致终止子获得的SNP位点个数；",
    "Missense Variant：样本在外显子区域的错义突变的SNP位点个数；",
    "Synonymous Variant：样本在外显子区域的同义突变的SNP位点个数；",
    "Intergenic Region：样本在基因间隔区的SNP位点个数。"
)
```

```{r snpannostat-table-html,eval=htmlfmt,results="asis"}
dt_caption("全基因组区域SNP位置分布信息统计表","snpannostat-table")
datatable(tempdf, rownames=FALSE,options = list(scrollX=T,columnDefs = list(list(className = 'dt-center',targets='_all'))))
dt_foot(tempfn)
```

```{r snpannostat-table,eval=pdffmt}
tempdf[,1]=gsub("_","\\\\_",tempdf[,1])
kbl(row.names=FALSE, tempdf, "latex", caption="全基因组区域SNP位置分布信息统计表", booktabs = TRUE, longtable = TRUE, linesep=' ',col.names=linebreak(gsub(" ","\n",colnames(tempdf)),align="c"),escape=FALSE) %>%
kable_styling(repeat_header_text="(续)", latex_options=c("striped","scale_down","repeat_header"))%>%
kbl_foot(tempfn) %>% table_bline()
```

```{r get-effectstat-table}
tempdf <- snp_anno_stat[,c("sampleID","HIGH","MODERATE","LOW","MODIFIER")]
colnames(tempdf) <- str_to_title(gsub("_"," ",colnames(tempdf)))
colnames(tempdf)[1] <- "Sample ID"
tempdf <- tempdf %>% tbl_prettier(c(NA,0,0,0,0))
tempdf <- tempdf[str_order(tempdf$`Sample ID`, numeric = T),]
tempfn <- c(
    "Sample ID：样品编号；",
    "High：具有破坏性影响，可能导致蛋白质功能丧失；",
    "Moderate：该类变异可能改变蛋白质的有效性；",
    "Low：该类变异大部分无害，不太可能改变蛋白质；",
    "Modifier：非编码变异或影响非编码基因的变异。"
)
```

```{r effectstat-table-html,eval=htmlfmt,results="asis"}
dt_caption("全基因组区域SNP功效信息统计表","effectstat-table")
datatable(tempdf, rownames=FALSE,options = list(scrollX=T,columnDefs = list(list(className = 'dt-center',targets='_all'))))
dt_foot(tempfn)
```

```{r effectstat-table,eval=pdffmt}
kbl(row.names=FALSE, tempdf, "latex", caption="全基因组区域SNP功效信息统计表", booktabs = TRUE, longtable = TRUE, linesep=' ') %>%
    kable_styling(repeat_header_text="(续)", latex_options=c("striped","scale_down","repeat_header"))%>%
    kbl_foot(tempfn) %>% table_bline()
```

### InDel检测

利用GATK的Best Practices流程处理比对结果（BAM文件），利用GATK的Haplotyper方法进行InDel检测及过滤，过滤条件按照GATK推荐的参数进行，具体可见：[https://software.broadinstitute.org/gatk/documentation/article.php?id=3225](https://software.broadinstitute.org/gatk/documentation/article.php?id=3225)。

对项目样品进行InDel标记开发，这里的InDel指能够明确获得序列组成的InDel标记。最终样本获得Insertion和Deletion详情统计结果如表\@ref(tab:indelstat-table)所示：

```{r get-indelstat-table}
indel_stat_file <- file.path(wgs_result_dir,"output","published","data", "04.snpIndel", "indel",  "indel.stat.xls")
arrange_indel_stat <- read.delim(indel_stat_file,sep='\t',header=TRUE,check.names = FALSE)[,1:5]
arrange_indel_stat <- arrange_indel_stat %>%
  tbl_prettier(c(NA, 0, 0, 0, 0))
arrange_indel_stat <- arrange_indel_stat[str_order(arrange_indel_stat$`Sample ID`, numeric = T),]
tempfn <- c(
"Sample ID：样品编号；",
"Insert Number：检测到的插入变异的位点个数；",
"Delete Number：检测到的缺失变异的位点个数；",
"Heterozygosity Number：杂合分型的InDel的位点个数；",
"Homozygosity Number：纯合分型的InDel位点个数。"
)
```

```{r indelstat-table-html,eval=htmlfmt,results="asis"}
dt_caption("InDel数据统计表","indelstat-table")
datatable(arrange_indel_stat, rownames=F,options = list(scrollX=T,columnDefs = list(list(className = 'dt-center',targets='_all'))))
dt_foot(tempfn)
```

```{r indelstat-table,eval=pdffmt}
arrange_indel_stat[,1] <- gsub("_","\\\\_",arrange_indel_stat[,1])
kbl(row.names=FALSE, arrange_indel_stat, "latex",caption="InDel数据统计表", booktabs = TRUE, longtable = TRUE, linesep=' ',col.names=linebreak(gsub(" ","\n",colnames(arrange_indel_stat)),align="c"),escape=FALSE) %>%
  kable_styling(repeat_header_text="(续)", latex_options=c("striped","scale_down","repeat_header"))%>%
  kbl_foot(tempfn) %>%
  table_bline()
```

### InDel功能注释

采用SnpEff程序结合本项目`r paste0(project_info[4],"（",project_info[5],"）")`基因组注释信息，对检测到的InDel进行功能注释，SnpEff会根据基因组的基因和功能区域的分布进行分析，对每个InDel所在的位置和功能进行统计，并对每个变异类型的功能进行统计。表\@ref(tab:indelannostat-table)为InDel位置分布信息统计表，表\@ref(tab:indeleffectstat-table)为InDel功效信息统计表：

> *完整的InDel功效信息统计表详见结题文件夹目录下[data_release/01.vcf_filter](./data_release/01.vcf_filter)文件夹下的pop.indel_anno.xls文件*

```{r get-indelannostat-table}
indel_anno_file <- file.path(wgs_result_dir,"output","published","data", "04.snpIndel", "indel", "indel_anno.stat.xls")
indel_anno <- read.delim(indel_anno_file, sep="\t", header=T, check.names=F)
arrange_indel_anno <- indel_anno %>%
  select(sampleID, frameshift_variant, intergenic_region, intragenic_variant,start_lost, stop_gained, stop_lost) %>%
  tbl_prettier(c(NA, 0, 0, 0, 0, 0, 0))
arrange_indel_anno <- arrange_indel_anno[str_order(arrange_indel_anno$sampleID, numeric = T),]
colnames(arrange_indel_anno)<-str_to_title(gsub("_"," ",colnames(arrange_indel_anno)))
colnames(arrange_indel_anno)[1] <- "Sample ID"
tempfn <- c(
    "Sample ID：样品编号；",
    "Frameshift Variant：导致移码突变的InDel个数",
    "Intergenic Region：在基因间区的InDel的个数；",
    "Intragenic Variant：在基因内非功能区的InDel个数；",
    "Start Lost：由于InDel的突变导致启动子缺失的InDel位点个数及所占比例；",
    "Stop Gained：由于InDel的突变导致终止子获得的InDel位点个数及所占比例；",
    "Stop Lost：由于InDel的突变导致终止子缺失的InDel位点个数及所占比例。"
)
```

```{r indelannostat-table-html,eval=htmlfmt,results="asis"}
dt_caption("全基因组区域InDel位置分布信息统计表","indelannostat-table")
datatable(arrange_indel_anno, rownames=FALSE,options = list(scrollX=T,columnDefs = list(list(className = 'dt-center',targets='_all'))))
dt_foot(tempfn)
```

```{r indelannostat-table,eval=pdffmt}
arrange_indel_anno[,1] <- gsub("_","\\\\_",arrange_indel_anno[,1])
kbl(row.names=FALSE, arrange_indel_anno, "latex", caption="全基因组区域InDel位置分布信息统计表", booktabs = TRUE, longtable = TRUE, linesep=' ',col.names=linebreak(gsub(" ","\n",colnames(arrange_indel_anno)),align="c"),escape=FALSE) %>%
  kable_styling(repeat_header_text="(续)", latex_options=c("striped","scale_down","repeat_header"))%>%
  kbl_foot(tempfn) %>%
  table_bline()
```

```{r get-indeleffectstat-table}
arrange_indel_impact <- indel_anno %>%
  select(sampleID, HIGH, MODERATE, LOW, MODIFIER) %>%
  tbl_prettier(c(NA, 0, 0, 0, 0))
colnames(tempdf) <- str_to_title(colnames(tempdf))
colnames(tempdf)[1] <- "Sample ID"
tempdf <- tempdf[str_order(tempdf$`Sample ID`, numeric = T),]
tempfn <- c(
    "Sample ID：样品编号；",
    "High：具有破坏性影响，可能导致蛋白质功能丧失；",
    "Moderate：该类变异可能改变蛋白质的有效性；",
    "Low：该类变异大部分无害，不太可能改变蛋白质；",
    "Modifier：非编码变异或影响非编码基因的变异。"
)
```

```{r indeleffectstat-table-html,eval=htmlfmt,results="asis"}
dt_caption("全基因组区域InDel功效信息统计表","indeleffectstat-table")
datatable(arrange_indel_impact, rownames=FALSE,options = list(scrollX=T,columnDefs = list(list(className = 'dt-center',targets='_all'))))
dt_foot(tempfn)
```

```{r indeleffectstat-table,eval=pdffmt}
kbl(row.names=FALSE, arrange_indel_impact, "latex", caption="全基因组区域InDel功效信息统计表", booktabs = TRUE, longtable = TRUE, linesep=' ') %>%
  kable_styling(repeat_header_text="(续)", latex_options=c("striped","scale_down","repeat_header"))%>%
  kbl_foot(tempfn) %>%
  table_bline()
```



